src/solvers/dg_p4est/containers_3d.jl

```diff
@@ -0,0 +1,303 @@
+# Initialize data structures in element container
+function init_elements!(elements, mesh::P4estMesh{3}, basis::LobattoLegendreBasis)
+  @unpack node_coordinates, jacobian_matrix,
+          contravariant_vectors, inverse_jacobian = elements
+
+  calc_node_coordinates!(node_coordinates, mesh, basis.nodes)
+
+  for element in 1:ncells(mesh)
+    calc_jacobian_matrix!(jacobian_matrix, element, node_coordinates, basis)
+
+    calc_contravariant_vectors!(contravariant_vectors, element, jacobian_matrix,
+                                node_coordinates, basis)
+
+    calc_inverse_jacobian!(inverse_jacobian, element, jacobian_matrix, basis)
+  end
+
+  return nothing
+end
+
+
+# Interpolate tree_node_coordinates to each quadrant
+function calc_node_coordinates!(node_coordinates,
+                                mesh::P4estMesh{3},
+                                basis::LobattoLegendreBasis)
+  # Hanging nodes will cause holes in the mesh if its polydeg is higher
+  # than the polydeg of the solver.
+  @assert length(basis.nodes) >= length(mesh.nodes) "The solver can't have a lower polydeg than the mesh"
+
+  calc_node_coordinates!(node_coordinates, mesh, basis.nodes)
+end
+
+function calc_node_coordinates!(node_coordinates,
```

Comment:
It would be nice to add some more commentary on what the purpose of the methods in this file is and also maybe here and there some comments inside the functions as to what is done (in order to guide future users to understand what's going on).